base/cvd/cuttlefish/host/libs/zip/zip_cc.h

```diff
@@ -0,0 +1,107 @@
+//
+// Copyright (C) 2025 The Android Open Source Project
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <stddef.h>
+#include <stdint.h>
+
+#include <memory>
+#include <string>
+
+#include "cuttlefish/common/libs/utils/result.h"
+
+namespace cuttlefish {
+
+class SeekableZipSourceCallback {
+ public:
+  virtual ~SeekableZipSourceCallback() = default;
+
+  virtual bool Close() = 0;
+  virtual bool Open() = 0;
+  virtual int64_t Read(char* data, uint64_t len) = 0;
+  virtual bool SetOffset(int64_t offset) = 0;
+  virtual int64_t Offset() = 0;
+  virtual uint64_t Size() = 0;
+};
+
+class ZipSource {
+ public:
+  friend class Zip;
+  friend class ZipSourceReader;
+
+  static Result<ZipSource> FromCallbacks(
+      std::unique_ptr<SeekableZipSourceCallback>);
+  static Result<ZipSource> FromData(const void* data, size_t size);
+  static Result<ZipSource> FromFile(const std::string& path);
+  static Result<ZipSource> FromZip(class Zip);
+
+  ZipSource(ZipSource&&);
+  ~ZipSource();
+  ZipSource& operator=(ZipSource&&);
+
+  Result<class ZipSourceReader> Reader();
```

Comment:
IIUC the only purpose of `ZipSource` is to get a `ZipSourceReader` out of it. Can the two classes be merged into one?

Reply:
There's a state change in the underlying `zip_source_t` that it can be "opened for read" or "opened for write", which are mutually exclusive operations so they can fail. The intention here is to tie the state change to the lifetime of the `ZipReader` instance, but I can see it being excessive at a time when the write operations are not exposed.